File: Lucy/Weekly Contest 405/3212_Count_Submatrices_With_Equal_Frequency_of_X_and_Y.js
/**
 * 3212. Count Submatrices With Equal Frequency of X and Y
 *
 * url: https://leetcode.com/problems/generate-binary-strings-without-adjacent-zeros/
 * difficulty: Medium
 * topic: Array, Matrix, Prefix Sum
 * date: 2024.07.15(MON)~
 */

/**
 * @param {character[][]} grid
 * @return {number}
 */
var numberOfSubmatrices = function (grid) {
  const GRID = Object.freeze({
    X: 1,
    Y: -1,
    ".": 0,
  });

  const isAtLeastOneX = (xCnt) => xCnt > 0;
  const isEqualFrequencyXAndY = (xCnt, yCnt) => xCnt - yCnt === 0;

  const calcPrefixSum = (prefixSum, grid) => {
    prefixSum[0][0] = GRID[grid[0][0]];

    for (let col = 1; col < prefixSum[0].length; col++) {
      prefixSum[0][col] = GRID[grid[0][col]] + prefixSum[0][col - 1];
    }

    for (let row = 1; row < prefixSum.length; row++) {
      prefixSum[row][0] = GRID[grid[row][0]] + prefixSum[row - 1][0];
    }

    for (let row = 1; row < prefixSum.length; row++) {
      for (let col = 1; col < prefixSum[row].length; col++) {
        prefixSum[row][col] =
          prefixSum[row - 1][col] +
          prefixSum[row][col - 1] -
          prefixSum[row - 1][col - 1] +
          GRID[grid[row][col]];
      }
    }
  };
  const calcSpecificCharPrefixSum = (prefixSum, grid, target) => {
    prefixSum[0][0] = grid[0][0] === target ? 1 : 0;

    for (let col = 1; col < prefixSum[0].length; col++) {
      prefixSum[0][col] =
        grid[0][col] === target
          ? prefixSum[0][col - 1] + 1
          : prefixSum[0][col - 1];
    }

    for (let row = 1; row < prefixSum.length; row++) {
      prefixSum[row][0] =
        grid[row][0] === target
          ? prefixSum[row - 1][0] + 1
          : prefixSum[row - 1][0];
    }

    for (let row = 1; row < prefixSum.length; row++) {
      for (let col = 1; col < prefixSum[row].length; col++) {
        prefixSum[row][col] =
          prefixSum[row - 1][col] +
          prefixSum[row][col - 1] -
          prefixSum[row - 1][col - 1] +
          (grid[row][col] === target ? 1 : 0);
      }
    }
  };

  const prefixSum = Array.from({ length: grid.length }, (_) =>
    new Array(grid[0].length).fill(0)
  );
  const xCntPrefixSum = Array.from({ length: grid.length }, (_) =>
    new Array(grid[0].length).fill(0)
  );
  const yCntPrefixSum = Array.from({ length: grid.length }, (_) =>
    new Array(grid[0].length).fill(0)
  );

  calcPrefixSum(prefixSum, grid);
  calcSpecificCharPrefixSum(xCntPrefixSum, grid, "X");
  calcSpecificCharPrefixSum(yCntPrefixSum, grid, "Y");

  let zeroCnt = 0;
  for (let i = 0; i < prefixSum.length; i++) {
    for (let j = 0; j < prefixSum[i].length; j++) {
      if (prefixSum[i][j] === 0) {
        if (
          isAtLeastOneX(xCntPrefixSum[i][j]) &&
          isEqualFrequencyXAndY(xCntPrefixSum[i][j], yCntPrefixSum[i][j])
        )
          zeroCnt++;
      }
    }
  }

  return zeroCnt;
};

console.log(
  numberOfSubmatrices([
    ["X", "Y", "."],
    ["Y", ".", "."],
  ])
);

File: Gombang/Weekly Contest 405/3212_Count_Submatrices_With_Equal_Frequency_of_X_and_Y.cs
public class Solution
{
    // Prefix Sum 알고리즘을 활용하여 진행.
    public int NumberOfSubmatrices(char[][] grid)
    {
        int result = 0;

        int n = grid.Length;
        int m = grid[0].Length;

        // x 및 y가 나오는 빈도수를 저장할 2차원 배열 두 개 생성.
        int[,] x = new int[n + 1, m + 1];
        int[,] y = new int[n + 1, m + 1];

        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                x[i + 1, j + 1] = x[i, j + 1] + x[i + 1, j] - x[i, j] + (grid[i][j] == 'X' ? 1 : 0);
                y[i + 1, j + 1] = y[i, j + 1] + y[i + 1, j] - y[i, j] + (grid[i][j] == 'Y' ? 1 : 0);

                // (i + 1, j + 1)에 해당하는 x의 빈도수 y빈도수가 동일하면 result값을 하나 증가.
                if (x[i + 1, j + 1] == y[i + 1, j + 1] && x[i + 1, j + 1] > 0)
                    result++;
            }
        }

        return result;
    }
}